lib/secretsscanner/scan/scan.go

```diff
@@ -0,0 +1,252 @@
+	file, err := os.Open(filePath)
+	if err != nil {
+		return nil, false, err
+	}
+	defer file.Close()
+
+	// read the first 150 bytes of the file to check if it's an OpenSSH private key.
+	var buf [150]byte
+	n, err := file.Read(buf[:])
+	if errors.Is(err, io.EOF) || n < len(buf) {
+		return nil, false, nil
+	} else if err != nil {
+		return nil, false, trace.Wrap(err, "failed to read file")
+	}
+
+	isPrivateKey := false
+	for _, header := range supportedPrivateKeyHeaders {
+		if bytes.Contains(buf[:], header) {
+			isPrivateKey = true
+			break
+		}
+	}
+	if !isPrivateKey {
+		return nil, false, nil
+	}
+
+	// read the entire file
+	data, err := io.ReadAll(file)
+	if err != nil {
+		return nil, false, trace.Wrap(err, "failed to read file")
+	}
+	return append(buf[:], data...), true, nil
+}
+
+func extractSSHKey(ctx context.Context, path, deviceID string, fileData []byte) (*accessgraphsecretsv1pb.PrivateKey, error) {
+	logger := slog.Default().With("private_key_file", path, "device_id", deviceID)
+
+	var publicKey ssh.PublicKey
+	var mode accessgraphsecretsv1pb.PublicKeyMode
+	var pme *ssh.PassphraseMissingError
+	switch pk, err := ssh.ParsePrivateKey(fileData); {
+	case errors.As(err, &pme):
+		if pme.PublicKey != nil {
+			// If the key is a OpenSSH private key whose public key is embedded in the header, it will return the public key. This is
+			// a special case for OpenSSH private keys that have the public key embedded in the header, for more information see
+			// OpenSSH's ssh key format: https://github.com/openssh/openssh-portable/blob/master/PROTOCOL.key
+			publicKey = pme.PublicKey
+			mode = accessgraphsecretsv1pb.PublicKeyMode_PUBLIC_KEY_MODE_DERIVED
+			break
+		}
+		publicKey, mode = parsePublicKeyFromPublicPath(ctx, logger, path)
+	case err != nil:
+		return nil, trace.Wrap(err)
+	default:
+		publicKey = pk.PublicKey()
+		mode = accessgraphsecretsv1pb.PublicKeyMode_PUBLIC_KEY_MODE_DERIVED
+	}
+	var fingerprint string
+	if publicKey != nil {
+		fingerprint = ssh.FingerprintSHA256(publicKey)
+	}
+
+	key, err := accessgraph.NewPrivateKeyWithName(
+		privateKeyNameGen(path, deviceID, fingerprint),
+		&accessgraphsecretsv1pb.PrivateKeySpec{
+			PublicKeyFingerprint: fingerprint,
+			DeviceId:             deviceID,
+			PublicKeyMode:        mode,
+		},
+	)
+	return key, trace.Wrap(err)
+}
+
+// parsePublicKeyFromPublicPath tries to read the public key from the public key file if the private key is password protected.
+// If the public key file doesn't exist, it will return mode accessgraphsecretsv1pb.PublicKeyMode_PUBLIC_KEY_MODE_PROTECTED
+// identifying that the private key is password protected and the public key could not be extracted.
+func parsePublicKeyFromPublicPath(ctx context.Context, logger *slog.Logger, path string) (publicKey ssh.PublicKey, mode accessgraphsecretsv1pb.PublicKeyMode) {
+	mode = accessgraphsecretsv1pb.PublicKeyMode_PUBLIC_KEY_MODE_PROTECTED
```

Comment:
Suggestion: Ditch the named returns and return in place?

`return nil, accessgraphsecretsv1pb.PublicKeyMode_PUBLIC_KEY_MODE_PROTECTED`

`return pub, accessgraphsecretsv1pb.PublicKeyMode_PUBLIC_KEY_MODE_PUB_FILE`

etc